### ISSUE TITLE
short_paths directory does not exist when build() gets called

### ISSUE THREAD

**Issue by monsdar** [conan-io/conan#541]
We're using the short_paths option on a number of our components because they have quite long names. Often we get the following error:

```
MyComponent/3.3.1@MyUser/thirdparty: Installing package bc88889b8e5ae166c1c71ad0e2b2ca98f4714a59
MyComponent/3.3.1@MyUser/thirCMake Error: The source directory "d:/.conan/tmp5vkq7x/MyComponent" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
dparty: Looking for package bc88889b8e5ae166c1c71ad0e2b2ca98f4714a59 in remote 'MyRemote' 
MyComponent/3.3.1@MyUser/thirdparty: WARN: Binary for bc88889b8e5ae166c1c71ad0e2b2ca98f4714a59 not in remote: 
MyComponent/3.3.1@MyUser/thirdparty: Building your package in d:/.conan\tmp5vkq7x
MyComponent/3.3.1@MyUser/thirdparty: Copying sources to build folder
MyComponent/3.3.1@MyUser/thirdparty: Generated cmake created conanbuildinfo.cmake
++++++++++++++++++++++THIS IS A PRINT FROM THE build() FUNCTION++++++++++++++
MyComponent/3.3.1@MyUser/thirdparty: ERROR: Package 'tmp5vkq7x' build failed
```

Notice how the CMake Error is splitting up the conan output. Also the output that the package gets build in `d:\.conan\...` comes later than the CMake error. If I look at `d:\.conan\` after all this is over I see the needed directory there (with just a conanbuildinfo.cmake in there).

It looks to me as if Conan trying to build the component before the directory has been created entirely.

I'm unsure how I could provide more information or how this could be reproduced... It does not happen everytime. Just wanted to report it as early as possible.

**Comment by memsharded** [comment]
Very weird, indeed. Can I see the `build()`` and ``source()`` methods? Maybe the full output too? If don't want to share private data, send an email, please. Thanks!

**Comment by monsdar** [comment]
We do not use `source()`, all the sources are included in the package.
Here's the `build()`:

```
def build(self):
    cmake = CMake(self.settings)
    
    GEOMETRY_ROOT = self.deps_cpp_info["ScoreGeometry"].rootpath
    GEOMETRY_ROOT = GEOMETRY_ROOT.replace('\\', '/')
    
    print "++++++++++++++++++++++" + GEOMETRY_ROOT
    self.run('cmake "%s/%s" %s -DCMAKE_INSTALL_PREFIX=install -DSCORE_GEOMETRY_INCLUDE_DIR=%s\include -DSCORE_GEOMETRY_LIBRARY_DIR=%s\lib -DSCORE_GEOMETRY_LIBRARY_STATIC=%s\lib -DSCORE_GEOMETRY_LIBRARY=%s\lib\score_geometry.lib -DSCORE_GEOMETRY_LIBRARY_STATIC=%s\lib\score_geometry-static.lib' % (self.conanfile_directory, self.name, cmake.command_line, GEOMETRY_ROOT, GEOMETRY_ROOT, GEOMETRY_ROOT, GEOMETRY_ROOT, GEOMETRY_ROOT))
    self.run('cmake --build . %s --target install' % cmake.build_config)
```

Please mind the dirty string substitution :)

**Comment by memsharded** [comment]
Certainly there is nothing in your build that could cause that behavior :(
From the paths it is clear that happens in Win, have you seen the same behavior in other OSs? Are you running the process concurrently or just a normal invocation?

**Comment by monsdar** [comment]
Just on Windows, our Linux build machine is a lot less used though. It's happening when running conan only once, no multitasking or similar.

**Comment by monsdar** [comment]
Had the same problem with the same lib on my local machine. Here's the output:

```
ScoreGeodesy/3.3.1@MyUser/thirdparty: WARN: Binary for bc88889b8e5ae166c1c71ad0e2b2ca98f4714a59 not in remote:
ScoreGeodesy/3.3.1@MyUser/thirdparty: Building your package in C:/.conan\tmpxflwr5
ScoreGeodesy/3.3.1@MyUser/thirdparty: Copying sources to build folder
ScoreGeodesy/3.3.1@MyUser/thirdparty: Generated cmake created conanbuildinfo.cmake
CMake Error: The source directory "C:/.conan/tmpxflwr5/ScoreGeodesy" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
++++++++++++++++++++++C:/Users/xxxxxxx/.conan/data/ScoreGeometry/2.3.5/MyUser/thirdparty/package/187b8d1a8b3c255d95cd824d7e9f892ea84c708d

ScoreGeodesy/3.3.1@MyUser/thirdparty: ERROR: Package 'tmpxflwr5' build failed
ScoreGeodesy/3.3.1@MyUser/thirdparty: WARN: Build folder C:/.conan\tmpxflwr5
```

I then deleted the content of `c:\.conan\` and the next build went through without problems. Running it again afterwards (without deleting the directory contents) worked flawlessly. I'm not sure why it was the Geodesy component again, it usually seems to be random.

**Comment by monsdar** [comment]
The error itself looks like Conan is creating the directory and continues working without working for the directory to be created. I had that some times with subprocess when I forgot to wait for it to finish:

```
subprocess.Popen("CreateDirectory").wait()
DoStuffWithTheCreatedDirectory()
```

Not that you're using subprocess for that, but perhaps using `os.mkdir` or something similar is not waiting for the directory to be created too.

**Comment by memsharded** [comment]
Thanks for the feedback, I'll try to run some more tests, I have tried a couple of things but still not able to replicate the error, I keep trying with your pointers.

**Comment by memsharded** [comment]
I have been doing some testing, launching hundreds of builds with short_paths=True, couldn't reproduce the issue :(
I am working in Win10, with python 2.7.11, 32 bits.

**Comment by monsdar** [comment]
We just had that problem again today :/
Win 7 Prof, Python 2.7.12 (I think), 64bit... We use Conan from Pip, perhaps we can put in some debug messages so we know what happens when it's occuring again?

**Comment by monsdar** [comment]
Just got a similar error that probably is related:

```
Foobar/1.0.22.4f6d90b@comp/release: Configuring sources in C:/.conan\tmpfalb6e
Traceback (most recent call last):
  File "C:\Python27\Scripts\conan-script.py", line 9, in <module>
    load_entry_point('conan==0.14.1', 'console_scripts', 'conan')()
  File "c:\python27\lib\site-packages\conans\conan.py", line 6, in run
    main(sys.argv[1:])
  File "c:\python27\lib\site-packages\conans\client\command.py", line 840, in main
    error = command.run(args)
  File "c:\python27\lib\site-packages\conans\client\command.py", line 739, in run
    method(args[0][1:])
  File "c:\python27\lib\site-packages\conans\client\command.py", line 383, in install
    profile_name=args.profile)
  File "c:\python27\lib\site-packages\conans\client\manager.py", line 298, in install
    installer.install(deps_graph, build_mode)
  File "c:\python27\lib\site-packages\conans\client\installer.py", line 59, in install
    self._build(nodes_by_level, skip_private_nodes, build_mode)
  File "c:\python27\lib\site-packages\conans\client\installer.py", line 139, in _build
    self._build_node(conan_ref, conan_file, build_mode)
  File "c:\python27\lib\site-packages\conans\client\installer.py", line 189, in _build_node
    conan_file, output)
  File "c:\python27\lib\site-packages\conans\client\installer.py", line 254, in _build_package
    config_source(export_folder, src_folder, conan_file, output)
  File "c:\python27\lib\site-packages\conans\client\source.py", line 41, in config_source
    shutil.copytree(export_folder, src_folder)
  File "C:\Python27\lib\shutil.py", line 177, in copytree
    os.makedirs(dst)
  File "C:\Python27\lib\os.py", line 157, in makedirs
    mkdir(name, mode)
WindowsError: [Error 183] Eine Datei kann nicht erstellt werden, wenn sie bereits vorhanden ist: 'C:/.conan\\tmpfalb6e'
```

Error message saying something on the line of `Cannot create file, it already exists`. Hope this helps.

**Comment by lasote** [comment]
It's related but not exactly the same problem.
Maybe the folder **C:/.conan\tmpfalb6e** is empty because of a bad execution with previous version.
Could you try to remove it and try again? 
Anyway It needs to be improved to not crash, so this is a bug, but please, try if it solves the problem.

Thanks!

**Comment by monsdar** [comment]
> Could you try to remove it and try again?

Removing the directory and retrying made the problem disappear..

**Comment by memsharded** [comment]
Great! Yes, the issue is that the new layout for short_paths is different than the old one, so it might need a conan cache clean before everything is fine. I guess that the original problem reported in this issue might persist, please report here if it happens again, we will try to introduce some logging and checks to narrow the bug search. Thanks!

**Comment by memsharded** [comment]
A question about the layout, and naming. I guess that every component is included in the corresponding recipe using ``exports`` under the same name of the package, so the layout of ``MyComponent`` will include a subfolder ``MyComponent`` that is exported into the recipe, is this right?

**Comment by monsdar** [comment]
Yeah that's right. It's essentially like this:

```
MyComponent
  - conanfile.py
  - MyComponent
  - - src
  - - CMakeLists.txt
```

**Comment by memsharded** [comment]
Unless @monsdar tell me the opposite, this feature is now fixed, will be released in next 0.15. 
Handling of path lenght limit with ``short_paths`` in Win has changed, it will require clearing the local cache. It has been merge to develop, if you want to try to run from the develop branch, to ensure that interactions with other changes do not affect it, and it is ready for 0.15 release.
